mypy/join.py

```diff
@@ -228,11 +228,13 @@ def visit_typeddict_type(self, t: TypedDictType) -> Type:
             items = OrderedDict([
                 (item_name, s_item_type)
                 for (item_name, s_item_type, t_item_type) in self.s.zip(t)
-                if is_equivalent(s_item_type, t_item_type)
+                if (is_equivalent(s_item_type, t_item_type) and
+                    (item_name in t.required_keys) == (item_name in self.s.required_keys))
             ])
             mapping_value_type = join_type_list(list(items.values()))
             fallback = self.s.create_anonymous_fallback(value_type=mapping_value_type)
-            return TypedDictType(items, fallback)
+            required_keys = set(items.keys()) & t.required_keys & self.s.required_keys
```

Comment:
Why is the `set(items.keys()) &` part needed? Is there ever a TypedDict whose required_keys is not a subset of its items?